File: go/04 space-age/space_age.go
/*
exercism.io go #4

Given an age in seconds, calculate how old someone would be on:

Earth: orbital period 365.25 Earth days, or 31557600 seconds
Mercury: orbital period 0.2408467 Earth years
Venus: orbital period 0.61519726 Earth years
Mars: orbital period 1.8808158 Earth years
Jupiter: orbital period 11.862615 Earth years
Saturn: orbital period 29.447498 Earth years
Uranus: orbital period 84.016846 Earth years
Neptune: orbital period 164.79132 Earth years

So if you were told someone were 1,000,000,000 seconds old, you should be able
to say that they're 31.69 Earth-years old.
*/
package space

type Planet = string

const secPerEarthyear = 31557600

// earth-year per planet-year
var periodMap = map[Planet]float64{
	"Earth":   1,
	"Jupiter": 11.8626215,
	"Mars":    1.8808158,
	"Mercury": 0.2408467,
	"Neptune": 164.79132,
	"Saturn":  29.447498,
	"Uranus":  84.016846,
	"Venus":   0.61519726,
}

func Age(sec float64, planet Planet) float64 {
	// sec * (earth-yr / sec) * (planet-yr / earth-yr) -> planet-yr
	return sec / secPerEarthyear / periodMap[planet]
}

File: scala/SpaceAge.scala
object SpaceAge {
  val mercury =   0.2408467
  val venus   =   0.61519726
  val earth   =   1
  val mars    =   1.8808158
  val jupiter =  11.862615
  val saturn  =  29.447498
  val uranus  =  84.016846
  val neptune = 164.79132

  val m: Double = 31557600 // sec / earth-year

  def foo(t: Double, p: Double) = (t * (1/m) * (1/p) * 100).round / 100.0

  def onEarth(t: Double)    = foo(t, earth)
  def onMercury(t: Double)  = foo(t, mercury)
  def onVenus(t: Double)    = foo(t, venus)
  def onMars(t: Double)     = foo(t, mars)
  def onJupiter(t: Double)  = foo(t, jupiter)
  def onSaturn(t: Double)   = foo(t, saturn)
  def onUranus(t: Double)   = foo(t, uranus)
  def onNeptune(t: Double)  = foo(t, neptune)
}